metadata/src/lib.rs

```diff
@@ -0,0 +1,458 @@
+                }
+            }
+            bytes.extend(&entry.default);
+        }
+    }
+
+    let pallet_hash = hash(&bytes);
+    cache.pallets.insert(pallet.name.clone(), pallet_hash);
+    pallet_hash
+}
+
+/// Obtain the hash representation of a `frame_metadata::RuntimeMetadataLastVersion`.
+pub fn get_metadata_hash(
+    metadata: &RuntimeMetadataLastVersion,
+    cache: &mut MetadataHasherCache,
+) -> [u8; 32] {
+    // Collect all pairs of (pallet name, pallet hash).
+    let mut pallets: Vec<(String, [u8; 32])> = metadata
+        .pallets
+        .iter()
+        .map(|pallet| {
+            let name = pallet.name.clone();
+            let hash = get_pallet_hash(&metadata.types, pallet, cache);
+            (name, hash)
+        })
+        .collect();
+    // Sort by pallet name to create a deterministic representation of the underlying metadata.
+    pallets.sort_by_key(|key| key.1);
+
+    // Note: pallet name is excluded from hashing.
+    // Each pallet has a hash of 32 bytes, and the vector is extended with
+    // extrinsic hash and metadata ty hash (2 * 32).
+    let mut bytes = Vec::with_capacity(pallets.len() * 32 + 64);
+    for (_, hash) in pallets.iter() {
+        bytes.extend(hash)
+    }
+
+    bytes.extend(get_extrinsic_hash(&metadata.types, &metadata.extrinsic));
+
+    let mut visited_ids = HashSet::<u32>::new();
+    bytes.extend(get_type_hash(
+        &metadata.types,
+        metadata.ty.id(),
+        &mut visited_ids,
+    ));
+
+    hash(&bytes)
+}
+
+/// Metadata hasher internal cache.
+#[derive(Clone, Debug)]
+pub struct MetadataHasherCache {
+    /// Cache of the pallets obtained from `get_pallet_hash`.
+    pub(crate) pallets: HashMap<String, [u8; 32]>,
+}
+
+impl MetadataHasherCache {
+    /// Creates an empty `MetadataHasherCache`.
+    pub fn new() -> Self {
+        Self {
+            pallets: HashMap::new(),
+        }
+    }
+}
+
+impl Default for MetadataHasherCache {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use crate::{
+        get_metadata_hash,
+        get_pallet_hash,
+        MetadataHasherCache,
+    };
+    use codec::Decode;
+    use frame_metadata::{
+        RuntimeMetadata::V14,
+        RuntimeMetadataLastVersion,
+        RuntimeMetadataPrefixed,
+    };
+    use scale_info::form::PortableForm;
+    use std::fs;
+
+    /// Metadata obtained from https://github.com/substrate-developer-hub/substrate-node-template.git
+    /// tag: polkadot-v0.9.17, branch origin/main.
+    static METADATA_PATH: &str = "./test-assets/node_template.scale";
+
+    /// Metadata obtained from https://github.com/substrate-developer-hub/substrate-node-template.git
+    /// tag: polkadot-v0.9.17, branch origin/main, via moving `Balances` pallet order as last
+    /// pallet in `construct_runtime` macro.
+    static METADATA_SWAP_PATH: &str = "./test-assets/node_template_swap.scale";
```

Comment:
One thing that you could think about for tests is manually creating `PalletMetadata` structs yourself, like we do here:

https://github.com/paritytech/subxt/blob/82f304005b700fabc6632f96f7945595f4da6673/subxt/src/events/events_type.rs#L444

We can also manually construct `PortableTypeRegistries`, eg like this:

https://github.com/paritytech/subxt/blob/82f304005b700fabc6632f96f7945595f4da6673/subxt/src/events/decoding.rs#L250

Being able to make those two things, it might become easier to have different tests which swap pallets around and play with different types and things (and which are more readable/updateable than the quite opaque `.scale` files), which would be really nice :)